### PR TITLE
Added mixed structured logging tests

### PR DESCRIPTION
**Pull request by 304NotModified**
See  #2532

### CODE REVIEW

File: tests/NLog.UnitTests/LoggerTests.cs [hunk 1]
```diff
@@ -1813,6 +1813,34 @@ public void SingleTargetMessageFormatOptimizationTest()
             AssertDebugLastMessage("target2", "DualTarget|Hello World");
         }
 
+        [Theory]
+        [InlineData(null, "OrderId", "@Client")]
+        [InlineData(true, "OrderId", "@Client")]
+        [InlineData(null, "0", "@Client")] //fails: gives "Process order {0} for {@Client}"
+        [InlineData(true, "0", "@Client")] //fails: gives "...ess order 13424 for 13424"
+        [InlineData(true, "OrderId", "Client")] //succeeeds, but gives JSON like (no quoted key, missing quotes arround string, =, other spacing)
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LoggerTests.cs]: not sure if this is expected, at least not the unquoted string

File: tests/NLog.UnitTests/LoggerTests.cs [hunk 2]
```diff
@@ -1813,6 +1813,34 @@ public void SingleTargetMessageFormatOptimizationTest()
             AssertDebugLastMessage("target2", "DualTarget|Hello World");
         }
 
+        [Theory]
+        [InlineData(null, "OrderId", "@Client")]
+        [InlineData(true, "OrderId", "@Client")]
+        [InlineData(null, "0", "@Client")] //fails: gives "Process order {0} for {@Client}"
+        [InlineData(true, "0", "@Client")] //fails: gives "...ess order 13424 for 13424"
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LoggerTests.cs]: bug for sure

File: tests/NLog.UnitTests/LoggerTests.cs [hunk 3]
```diff
@@ -1813,6 +1813,34 @@ public void SingleTargetMessageFormatOptimizationTest()
             AssertDebugLastMessage("target2", "DualTarget|Hello World");
         }
 
+        [Theory]
+        [InlineData(null, "OrderId", "@Client")]
+        [InlineData(true, "OrderId", "@Client")]
+        [InlineData(null, "0", "@Client")] //fails: gives "Process order {0} for {@Client}"
```
Comment (304NotModified) [inline tests/NLog.UnitTests/LoggerTests.cs]: Doubting if this is a bug, at least it's unwanted

**Comment by snakefoot** [comment]
I agree that parseMessageTemplate = true should work. Please also try with the parameters reverted.



Sent from my Samsung device


-------- Original message --------

File: tests/NLog.UnitTests/LoggerTests.cs (hunk 3 above)
Comment (snakefoot) [reply tests/NLog.UnitTests/LoggerTests.cs]: Works as expected, not valid string.Format input, so it just returns the raw unformatted message. Unless you want to have fallback on exception?

**Comment by 304NotModified** [comment]
@snakefoot I hacked the change, but now I have removed the support for index holes in a structured template. 

That's wrong, I need to fix that, but that's a bit hard because the parsing is lazy.

**Comment by snakefoot** [comment]
> now I have removed the support for index holes in a structured template.

Thats sounds worse than not supporting mixed templates.

**Comment by 304NotModified** [comment]
> You have also introduced a nice performance overhead, by always allocating message-parameter-objects for index-holes.

that's because of `var currentHole = holeEnumerator.Current.Hole;`?

**Comment by 304NotModified** [comment]
superseded by #2544

**Pull request closed without merging**